// node-disk-storage/dist/lib/index.js
/**
 * node-disk-storage
 * @author Copyright(c) 2021 by Restu wahyu saputra
 * MIT Licensed
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "is-any-type", "../utils/storage", "../utils/validator"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.NDS = void 0;
    const is_any_type_1 = require("is-any-type");
    const store = __importStar(require("../utils/storage"));
    const validator_1 = require("../utils/validator");
    class NDS {
        minSize;
        maxSize;
        compress;
        items = [];
        options;
        constructor(options = {}) {
            this.options = options;
            this.minSize = options.minSize || 1;
            this.maxSize = options.maxSize || 26;
            this.compress = options.compress || false;
        }
        /**
         * set data using key and value, into disk
         *
         * @param { String } input - required
         * @param { string } value - required
         * @return boolean | undefined
         */
        set(key, value) {
            if (is_any_type_1.assert.isBoolean(validator_1.validatorKeyVal({ key, value }))) {
                const options = {
                    minSize: this.minSize,
                    maxSize: this.maxSize,
                    compress: this.compress
                };
                if (is_any_type_1.assert.isBoolean(validator_1.sizeValidator(options, value))) {
                    this.items.push({ key, value });
                    return store.setItem(this.items, options.compress, this.options);
                }
            }
        }
        /**
         * get specific data using key, after saving data into disk
         *
         * @param { String } key - required
         * @return string | undefined
         */
        get(key) {
            if (is_any_type_1.assert.isBoolean(validator_1.validatorKey(key))) {
                return store.getItem(key, this.compress, this.options);
            }
        }
        /**
         * remove specific data already exist using key, after saving data into disk
         *
         * @param { String } key - required
         * @return boolean | undefined
         */
        remove(key) {
            if (is_any_type_1.assert.isBoolean(validator_1.validatorKey(key))) {
                return store.removeItem(key, this.options);
            }
        }
        /**
         * clear all keys exist, after saving data into disk
         *
         * @return boolean | undefined
         */
        clear() {
            return store.clearItem(this.options);
        }
        /**
         * get all keys exist, after saving data into disk
         *
         * @return string[] | undefined
         */
        keys() {
            return store.keysItem(this.compress, this.options);
        }
    }
    exports.NDS = NDS;
});
