// node-disk-storage/dist/utils/compress.js
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "zlib"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.unzip = exports.zip = void 0;
    const zlib_1 = __importDefault(require("zlib"));
    const zip = (items) => {
        return zlib_1.default.gzipSync(Buffer.from(JSON.stringify({ items })), {
            level: zlib_1.default.constants.Z_BEST_COMPRESSION,
            memLevel: zlib_1.default.constants.Z_RLE,
            strategy: zlib_1.default.constants.Z_RLE,
            chunkSize: 999999999,
            maxOutputLength: 999999999
        });
    };
    exports.zip = zip;
    const unzip = (items) => {
        const data = zlib_1.default
            .gunzipSync(items, {
            level: zlib_1.default.constants.Z_BEST_COMPRESSION,
            memLevel: zlib_1.default.constants.Z_RLE,
            strategy: zlib_1.default.constants.Z_RLE,
            chunkSize: 999999999,
            maxOutputLength: 999999999
        })
            .toString('utf-8');
        return JSON.parse(data).items;
    };
    exports.unzip = unzip;
});
